### ISSUE TITLE
Implement CachedBeaconState

### ISSUE THREAD

**Issue by twoeths** [ChainSafe/lodestar#1950]
The idea came from https://github.com/ChainSafe/lodestar/pull/1925#pullrequestreview-564581634

> Going forward, we should take some time to possibly re-organize the beacon state, validator cache, epoch context, epoch process, etc. as needed.

> This is mainly an issue of how we organize the various datastructures (BeaconState, CachedValidatorsBeaconState, EpochContext, EpochProcess, IStateContext). I'm not sure what the right organization is (possibly no re-organization needed) but what we have seems to be getting more cluttered as we've added more and more optimizations. We have several nested datastructures that are all related to the beacon state / epoch state.

> My hunch is that we may be able to expand the CachedValidatorBeaconState to be more of a general CachedBeaconState at the 'top-level', like IStateContext, which exposes the BeaconState + extra methods, with various caches 'under the hood' (each cache using 'persistent-ts' or similar for cheap cloning).

Need to come to an agreement regarding how it looks like before implementation.

**Comment by twoeths** [comment]
one good thing with caching IFlatValidator (which is currently in CachedValidatorsBeaconState) is we can pull `epochBalances` directly from there instead of having to publish `epochProcess` (in lodestar-beacon-state-transition) to lodestar, the consequence of that is:
+ No need different types of EpochContext (StateTransitionEpochContext, LodestarEpochContext), only EpochContext
+ `epochProcess` stays internal to lodestar-beacon-state-transition and hence it's not a concern

we need to do this cleanup work before implementing CachedBeaconState

**Comment by twoeths** [comment]
so the class looks like this
```ts
export class CachedBeaconState extends EpochContext {
  ...// old stuff of CachedValidatorsBeaconState
  // same method to EpochContext without param
  public rotateEpochs(): void {
    super.rotateEpochs(this);
  }
  // same method to EpochContext without param
  public syncPubkeys(): void {
    super.syncPubkeys(this.getOriginalState());
  }
}
```

+ we separate EpochContext (epoch stuff cache) and other BeaconState related cache while consumer only know CacheBeaconState and it has everything it needs
+ `CachedBeaconState.clone()` is cheap
+ EpochContext becomes abstract, noone should use it directly
+ stateTransition function looks like:
```ts
export function fastStateTransition(
  cachedState: CachedBeaconState,
  signedBlock: SignedBeaconBlock,
  options?: {verifyStateRoot?: boolean; verifyProposer?: boolean; verifySignatures?: boolean}
): CachedBeaconState
```
+ lodestar should use this new CachedBeaconState across the package instead of IStateContext/ITreeStateContext which is removed

**Comment by wemeetagain** [comment]
resolved in #2172
